### ISSUE TITLE
Debian packaging of crazy-complete

### ISSUE THREAD

**Issue by n-peugnet** [crazy-complete/crazy-complete#24]
Hi, I am currently in the process of packaging crazy-complete for Debian and I have a few questions and requests.

1.  I noticed the LICENSE file of the GPLv3.0, but you did not explicitly indicate if you allow to use later versions. This is usually done by adding the following notice at the top of your files, and keeping it as is or modifying the `either version 3 of the License, or (at your option) any later version` part:

    ```
    <one line to give the program's name and a brief idea of what it does.>
    Copyright (C) <year>  <name of author>

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.
    ```

    I would suggest to add this notice at least on the main `crazy-complete` file.

2.  Related to 1. I didn't find an explicit copyright statement in your files, it would be good to include one in this same notice. You can also add an email address so that one can contact you.
3.  As I noticed that there was quite a lot of changes between 0.3.2 and current main, I decided to start packaging from the main branch, but ideally I would need a properly tagged release. Do you intend to make a release soon?

Apart from this, the package is almost ready, you can take a look at my packaging work on [Salsa](https://salsa.debian.org/n-peugnet/crazy-complete/-/tree/debian/latest/debian?ref_type=heads) (I used a temporary version that will be replaced by the proper one once it will be released).

**Comment by n-peugnet** [comment]
Hi @braph, please take a look at this issue if you have some time.

**Comment by braph** [comment]
Hello @n-peugnet ,

I reread your issue from the beginning.

Regarding the GPL version:

Should I add `# SPDX-License-Identifier: GPL-3.0-or-later` to all source files?

And should I include the following notice in the `crazy-complete` script?

```
crazy-complete - Generate shell auto completion files
Copyright (C) 2025-2026 Benjamin Abendroth (braph93@gmx.de)

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
GNU General Public License for more details.
```

I also have a question for you: do you use the GitHub tarballs or the mirrored/forked sources from
https://salsa.debian.org/debian/crazy-complete/ to build the packages?

The current official release on GitHub is already somewhat outdated. I would create a new release if needed (version 0.3.7).

The sources on salsa.debian.org are also not up to date.

It is important to me that the latest release is packaged.

Only a few bugs were fixed, but many new features were added.

I'm not sure how familiar you are with crazy-complete, so here is a shortened list of new features:

- Many new built-in completers were added. The most useful one is probably `command_arg`, which allows sudo-like behavior (completing a command and its arguments).

- New meta completers were added (`list`, `key_value_list`), which allow combining arbitrary completers.
  For example, `complete: ["list", ["combine", [["user", "uid"]]]]` would complete a list of usernames or user IDs.

- A new `wraps` option was added to wrap another command and reuse its completion behavior.

A complete list of new features would be included in the release notes.

Regarding the issue with different CPU architectures: I will set up some QEMU VMs (PPC, RISC-V, s390x) over the next couple of days and try to reproduce the problem. Based on the logs, it does not look like a timing issue. I will only be able to say what exactly is happening once I have analyzed it myself.

Kind regards,

Benjamin

**Comment by braph** [comment]
Hello @n-peugnet ,

I added the SPDX-License-Identifier including my name and email address to all source files.

I also added the long version of the license header to the main crazy-complete script. The project is licensed under GPL-3.0-or-later, so I allow any version of GPL-3 or later.

Regarding Debian packaging: I don't think it is necessary to include crazy-complete in stable-backports. That would just create additional work for you, and users can always install the latest version from GitHub if they need it.

I also investigated the problems occurring during testing on RISC-V. It turned out to be a timing issue. I solved it by increasing the wait timeout in the tests to 5 seconds. The main difficulty when testing tab completion is that there is no reliable way to detect when the completion output is finished, since it is just a stream of text without any marker indicating end of completion.

After updating my system, I encountered another issue: Fish shell now appears to use a looser completion behavior. For example, `foo -F` may also suggest `foo --flag` as a completion. Because of this, two tests fail on my system now. However, the tests should still pass on your side until Fish is updated to **4.4** (I could not find any mention of this completion behavior change in the Fish release notes).

I am currently trying to find a good solution for testing with different Fish versions.

Sorry for the delay in responding - another user requested a feature that I had to implement first.

Regards

Benjamin

**Comment by n-peugnet** [comment]
> I added the SPDX-License-Identifier including my name and email address to all source files.
> 
> I also added the long version of the license header to the main crazy-complete script. The project is licensed under GPL-3.0-or-later, so I allow any version of GPL-3 or later.

Nice!

> Regarding Debian packaging: I don't think it is necessary to include crazy-complete in stable-backports. That would just create additional work for you, and users can always install the latest version from GitHub if they need it.

I will see if any user requests it, I might still do it. As crazy-complete depends on very little other packages it would be quite easy to backport.

> I also investigated the problems occurring during testing on RISC-V. It turned out to be a timing issue. I solved it by increasing the wait timeout in the tests to 5 seconds. The main difficulty when testing tab completion is that there is no reliable way to detect when the completion output is finished, since it is just a stream of text without any marker indicating end of completion.

I understand, I decided to run again the tests on all the architectures supported by Debian, I re-enabled the `--threads` option and used a timeout of 5 seconds. Let's see how it behaves now.

> After updating my system, I encountered another issue: Fish shell now appears to use a looser completion behavior. For example, `foo -F` may also suggest `foo --flag` as a completion. Because of this, two tests fail on my system now. However, the tests should still pass on your side until Fish is updated to **4.4** (I could not find any mention of this completion behavior change in the Fish release notes).
> 
> I am currently trying to find a good solution for testing with different Fish versions.

Thank you for letting me know. Maybe an easy workaround that could be done quickly would be to allow selecting a subset of the supported shells when running the tests, this way I could run the test only for bash and zsh and disable them for fish while they are not compatible with our fish version? Some flag like `--shells=bash,zsh` would do the job.

**Comment by n-peugnet** [comment]
> > The main difficulty when testing tab completion is that there is no reliable way to detect when the completion output is finished, since it is just a stream of text without any marker indicating end of completion.
> 
> I understand, I decided to run again the tests on all the architectures supported by Debian, I re-enabled the `--threads` option and used a timeout of 5 seconds. Let's see how it behaves now.

Well it seems 5s was not enough on some of our (overloaded) machines, so instead, I decided to use the `--fast` option together with a very long `--wait=20`. From what I understand of the code, the `--fast` option effectively turns the `--wait` option into a timeout, rather than a time we have to wait on each test:

https://github.com/crazy-complete/crazy-complete/blob/40d4917296647fb9eaf82c5a6f66b473c1b51888/test/tests/terminal_base.py#L32-L41

This makes it the best solution to make the tests run fast on fast systems, while allowing them to take the time needed on slower ones.

By the way, I find the description of the `--fast` option a bit misleading:
> Makes the tests run even faster at the cost of corectness for some tests. For tests where the input matches the expected output, these tests will always pass.

From what I understand, given the same `--wait` value, no more tests should fail with the `--fast` option than without, right? As a fast test will only return faster if the output is correct, or else, it will wait the same amount of time as the non-fast test, and return the same result.

**Comment by n-peugnet** [comment]
> By the way, I find the description of the `--fast` option a bit misleading:
> 
> > Makes the tests run even faster at the cost of corectness for some tests. For tests where the input matches the expected output, these tests will always pass.
> 
> From what I understand, given the same `--wait` value, no more tests should fail with the `--fast` option than without, right? As a fast test will only return faster if the output is correct, or else, it will wait the same amount of time as the non-fast test, and return the same result.

Ah no OK, I got it, the risk is not to get false-positives, but rather to get false-negative (i.e. getting no test failure when we should have had one), especially for tests that expect no additional output from the shell.

But I think this is actually a good trade-off for Debian, as it is especially important for us to not get false-positives.

I will close this issue as everything is now good for me. BTW, I haven't had any new test failures with the patches I sent you :)
